x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialContains.java

```diff
@@ -0,0 +1,280 @@
+ * The bulk of the capabilities are within the parent class SpatialRelatesFunction,
+ * which supports all the relations in the ShapeField.QueryRelation enum.
+ * Here we simply wire the rules together specific to ST_CONTAINS and QueryRelation.CONTAINS.
+ */
+public class SpatialContains extends SpatialRelatesFunction {
+    // public for test access with reflection
+    public static final SpatialRelationsContains GEO = new SpatialRelationsContains(
+        SpatialCoordinateTypes.GEO,
+        CoordinateEncoder.GEO,
+        new GeoShapeIndexer(Orientation.CCW, "ST_Contains")
+    );
+    // public for test access with reflection
+    public static final SpatialRelationsContains CARTESIAN = new SpatialRelationsContains(
+        SpatialCoordinateTypes.CARTESIAN,
+        CoordinateEncoder.CARTESIAN,
+        new CartesianShapeIndexer("ST_Contains")
+    );
+
+    /**
+     * We override the normal behaviour for CONTAINS because we need to test each component separately.
+     * This applies to multi-component geometries (MultiPolygon, etc.) as well as polygons that cross the dateline.
+     */
+    static final class SpatialRelationsContains extends SpatialRelations {
+        SpatialRelationsContains(SpatialCoordinateTypes spatialCoordinateType, CoordinateEncoder encoder, ShapeIndexer shapeIndexer) {
+            super(ShapeField.QueryRelation.CONTAINS, spatialCoordinateType, encoder, shapeIndexer);
+        }
+
+        @Override
+        protected boolean geometryRelatesGeometry(BytesRef left, BytesRef right) throws IOException {
+            Component2D[] rightComponent2Ds = asLuceneComponent2Ds(crsType, fromBytesRef(right));
+            return geometryRelatesGeometries(left, rightComponent2Ds);
+        }
+
+        private boolean geometryRelatesGeometries(BytesRef left, Component2D[] rightComponent2Ds) throws IOException {
+            Geometry leftGeom = fromBytesRef(left);
+            GeometryDocValueReader leftDocValueReader = asGeometryDocValueReader(coordinateEncoder, shapeIndexer, leftGeom);
+            return geometryRelatesGeometries(leftDocValueReader, rightComponent2Ds);
+        }
+
+        private boolean geometryRelatesGeometries(GeometryDocValueReader leftDocValueReader, Component2D[] rightComponent2Ds)
+            throws IOException {
+            for (Component2D rightComponent2D : rightComponent2Ds) {
+                // Every component of the right geometry must be contained within the left geometry for this to pass
+                if (geometryRelatesGeometry(leftDocValueReader, rightComponent2D) == false) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    @FunctionInfo(returnType = { "boolean" }, description = "Returns whether the first geometry contains the second geometry.")
+    public SpatialContains(
+        Source source,
+        @Param(
+            name = "geomA",
```

Comment:
I'd use a full word. This will be used by Kibana for autocomplete, but I am not sure if as is or they will give it a proper name.

Reply:
I used `geomA` and `geomB` because that is exactly what PostGIS used, and I kind-of liked the idea that our docs look a little similar to PostGIS docs. But Kibana obviously takes precedence. I could ask them about this.